osu.Game/Screens/Select/BeatmapInfoWedge.cs

```diff
@@ -383,15 +383,24 @@ private InfoLabel[] getInfoLabels()
                 return labels.ToArray();
             }
 
+            [Resolved]
+            private IBindable<IReadOnlyList<Mod>> mods { get; set; }
```

Comment:
This is pretty weird. The class is receiving in a beatmap + ruleset, but then DI-ing in the mods.

It also doesn't respond to changes in mod settings, which are a little more complex to handle in the context of this class SEVERELY needing a rewrite. This is the sort of change required to fix both of these issues:

```diff
diff --git a/osu.Game/Screens/Select/BeatmapInfoWedge.cs b/osu.Game/Screens/Select/BeatmapInfoWedge.cs
index 311ed6ffb9..08a97b32c1 100644
--- a/osu.Game/Screens/Select/BeatmapInfoWedge.cs
+++ b/osu.Game/Screens/Select/BeatmapInfoWedge.cs
@@ -5,7 +5,6 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
-using JetBrains.Annotations;
 using osuTK;
 using osuTK.Graphics;
 using osu.Framework.Allocation;
@@ -25,6 +24,7 @@
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Localisation;
 using osu.Framework.Logging;
+using osu.Game.Configuration;
 using osu.Game.Rulesets;
 using osu.Game.Rulesets.Mods;
 using osu.Game.Rulesets.UI;
@@ -38,7 +38,11 @@ public class BeatmapInfoWedge : VisibilityContainer
 
         private static readonly Vector2 wedged_container_shear = new Vector2(shear_width / SongSelect.WEDGE_HEIGHT, 0);
 
-        private readonly IBindable<RulesetInfo> ruleset = new Bindable<RulesetInfo>();
+        [Resolved]
+        private IBindable<RulesetInfo> ruleset { get; set; }
+
+        [Resolved]
+        private IBindable<IReadOnlyList<Mod>> mods { get; set; }
 
         [Resolved]
         private BeatmapDifficultyCache difficultyCache { get; set; }
@@ -64,9 +68,8 @@ public BeatmapInfoWedge()
         }
 
         [BackgroundDependencyLoader(true)]
-        private void load([CanBeNull] Bindable<RulesetInfo> parentRuleset)
+        private void load()
         {
-            ruleset.BindTo(parentRuleset);
             ruleset.ValueChanged += _ => updateDisplay();
         }
 
@@ -132,7 +135,7 @@ void removeOldInfo()
                     return;
                 }
 
-                LoadComponentAsync(loadingInfo = new BufferedWedgeInfo(beatmap, ruleset.Value, beatmapDifficulty.Value)
+                LoadComponentAsync(loadingInfo = new BufferedWedgeInfo(beatmap, ruleset.Value, mods.Value, beatmapDifficulty.Value)
                 {
                     Shear = -Shear,
                     Depth = Info?.Depth + 1 ?? 0
@@ -164,16 +167,21 @@ public class BufferedWedgeInfo : BufferedContainer
 
             private ILocalisedBindableString titleBinding;
             private ILocalisedBindableString artistBinding;
+            private Container bpmLabelContainer;
 
             private readonly WorkingBeatmap beatmap;
             private readonly RulesetInfo ruleset;
+            private readonly IReadOnlyList<Mod> mods;
             private readonly StarDifficulty starDifficulty;
 
-            public BufferedWedgeInfo(WorkingBeatmap beatmap, RulesetInfo userRuleset, StarDifficulty difficulty)
+            private ModSettingChangeTracker settingChangeTracker;
+
+            public BufferedWedgeInfo(WorkingBeatmap beatmap, RulesetInfo userRuleset, IReadOnlyList<Mod> mods, StarDifficulty difficulty)
                 : base(pixelSnapping: true)
             {
                 this.beatmap = beatmap;
                 ruleset = userRuleset ?? beatmap.BeatmapInfo.Ruleset;
+                this.mods = mods;
                 starDifficulty = difficulty;
             }
 
@@ -184,9 +192,11 @@ private void load(LocalisationManager localisation)
                 var metadata = beatmapInfo.Metadata ?? beatmap.BeatmapSetInfo?.Metadata ?? new BeatmapMetadata();
 
                 CacheDrawnFrameBuffer = true;
-
                 RelativeSizeAxes = Axes.Both;
 
+                settingChangeTracker = new ModSettingChangeTracker(mods);
+                settingChangeTracker.SettingChanged += _ => updateBPM();
+
                 titleBinding = localisation.GetLocalisedString(new LocalisedString((metadata.TitleUnicode, metadata.Title)));
                 artistBinding = localisation.GetLocalisedString(new LocalisedString((metadata.ArtistUnicode, metadata.Artist)));
 
@@ -307,7 +317,25 @@ private void load(LocalisationManager localisation)
                                 Margin = new MarginPadding { Top = 20 },
                                 Spacing = new Vector2(20, 0),
                                 AutoSizeAxes = Axes.Both,
-                                Children = getInfoLabels()
+                                Children = new Drawable[]
+                                {
+                                    new InfoLabel(new BeatmapStatistic
+                                    {
+                                        Name = "Length",
+                                        CreateIcon = () => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Length),
+                                        Content = TimeSpan.FromMilliseconds(beatmapInfo.Length).ToString(@"m\:ss"),
+                                    }),
+                                    bpmLabelContainer = new Container
+                                    {
+                                        AutoSizeAxes = Axes.Both,
+                                    },
+                                    new FillFlowContainer
+                                    {
+                                        AutoSizeAxes = Axes.Both,
+                                        Spacing = new Vector2(20, 0),
+                                        Children = getRulesetInfoLabels()
+                                    }
+                                }
                             }
                         }
                     }
@@ -319,6 +347,8 @@ private void load(LocalisationManager localisation)
                 // no difficulty means it can't have a status to show
                 if (beatmapInfo.Version == null)
                     StatusPill.Hide();
+
+                updateBPM();
             }
 
             private static Drawable createStarRatingDisplay(StarDifficulty difficulty) => difficulty.Stars > 0
@@ -335,72 +365,60 @@ private void setMetadata(string source)
                 ForceRedraw();
             }
 
-            private InfoLabel[] getInfoLabels()
+            private InfoLabel[] getRulesetInfoLabels()
             {
-                var b = beatmap.Beatmap;
-
-                List<InfoLabel> labels = new List<InfoLabel>();
-
-                if (b?.HitObjects?.Any() == true)
+                try
                 {
-                    labels.Add(new InfoLabel(new BeatmapStatistic
-                    {
-                        Name = "Length",
-                        CreateIcon = () => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Length),
-                        Content = TimeSpan.FromMilliseconds(b.BeatmapInfo.Length).ToString(@"m\:ss"),
-                    }));
-
-                    labels.Add(new InfoLabel(new BeatmapStatistic
-                    {
-                        Name = "BPM",
-                        CreateIcon = () => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Bpm),
-                        Content = getBPMRange(b),
-                    }));
+                    IBeatmap playableBeatmap;
 
                     try
                     {
-                        IBeatmap playableBeatmap;
-
-                        try
-                        {
-                            // Try to get the beatmap with the user's ruleset
-                            playableBeatmap = beatmap.GetPlayableBeatmap(ruleset, Array.Empty<Mod>());
-                        }
-                        catch (BeatmapInvalidForRulesetException)
-                        {
-                            // Can't be converted to the user's ruleset, so use the beatmap's own ruleset
-                            playableBeatmap = beatmap.GetPlayableBeatmap(beatmap.BeatmapInfo.Ruleset, Array.Empty<Mod>());
-                        }
-
-                        labels.AddRange(playableBeatmap.GetStatistics().Select(s => new InfoLabel(s)));
+                        // Try to get the beatmap with the user's ruleset
+                        playableBeatmap = beatmap.GetPlayableBeatmap(ruleset, Array.Empty<Mod>());
                     }
-                    catch (Exception e)
+                    catch (BeatmapInvalidForRulesetException)
                     {
-                        Logger.Error(e, "Could not load beatmap successfully!");
+                        // Can't be converted to the user's ruleset, so use the beatmap's own ruleset
+                        playableBeatmap = beatmap.GetPlayableBeatmap(beatmap.BeatmapInfo.Ruleset, Array.Empty<Mod>());
                     }
+
+                    return playableBeatmap.GetStatistics().Select(s => new InfoLabel(s)).ToArray();
+                }
+                catch (Exception e)
+                {
+                    Logger.Error(e, "Could not load beatmap successfully!");
                 }
 
-                return labels.ToArray();
+                return Array.Empty<InfoLabel>();
             }
 
-            [Resolved]
-            private IBindable<IReadOnlyList<Mod>> mods { get; set; }
-
-            private string getBPMRange(IBeatmap beatmap)
+            private void updateBPM()
             {
+                var b = beatmap.Beatmap;
+                if (b == null)
+                    return;
+
                 // this doesn't consider mods which apply variable rates, yet.
                 double rate = 1;
-                foreach (var mod in mods.Value.OfType<IApplicableToRate>())
+                foreach (var mod in mods.OfType<IApplicableToRate>())
                     rate = mod.ApplyToRate(0, rate);
 
-                double bpmMax = beatmap.ControlPointInfo.BPMMaximum * rate;
-                double bpmMin = beatmap.ControlPointInfo.BPMMinimum * rate;
-                double mostCommonBPM = 60000 / beatmap.GetMostCommonBeatLength() * rate;
+                double bpmMax = b.ControlPointInfo.BPMMaximum * rate;
+                double bpmMin = b.ControlPointInfo.BPMMinimum * rate;
+                double mostCommonBPM = 60000 / b.GetMostCommonBeatLength() * rate;
+
+                string labelText = Precision.AlmostEquals(bpmMin, bpmMax)
+                    ? $"{bpmMin:0}"
+                    : $"{bpmMin:0}-{bpmMax:0} (mostly {mostCommonBPM:0})";
 
-                if (Precision.AlmostEquals(bpmMin, bpmMax))
-                    return $"{bpmMin:0}";
+                bpmLabelContainer.Child = new InfoLabel(new BeatmapStatistic
+                {
+                    Name = "BPM",
+                    CreateIcon = () => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Bpm),
+                    Content = labelText
+                });
 
-                return $"{bpmMin:0}-{bpmMax:0} (mostly {mostCommonBPM:0})";
+                ForceRedraw();
             }
 
             private OsuSpriteText[] getMapper(BeatmapMetadata metadata)
@@ -423,6 +441,12 @@ private OsuSpriteText[] getMapper(BeatmapMetadata metadata)
                 };
             }
 
+            protected override void Dispose(bool isDisposing)
+            {
+                base.Dispose(isDisposing);
+                settingChangeTracker?.Dispose();
+            }
+
             public class InfoLabel : Container, IHasTooltip
             {
                 public string TooltipText { get; }
```

Reply:
I've committed the above diff, with a slight change due to the parent class' weird DI-ing. Please check!